### ISSUE TITLE
Application crash when using Ammo

### ISSUE THREAD

**Issue by LeXXik** [playcanvas/engine#6723]
An application crashes, when using Ammo. This is due to setting a JS Function in Wasm, which expects a WebAssembly.Function:

https://github.com/playcanvas/engine/blob/eba06683346ede7a2932c5e58fbff85a0308f33a/src/framework/components/rigid-body/system.js#L386-L388

To repro:
- open engine examples, Vehicle
- place a breakpoint at above line in the rigidbody system
- refresh and once stopped at breakpoint, enable pause on caught exceptions
- resume execution

![image](https://github.com/playcanvas/engine/assets/5677782/c5afd914-fd28-473d-af0a-515fc0c17977)

Unfortunately, WebAssembly.Function is in stage 3 proposal today, and cannot be used now without browsers using special flags. There is a workaround of compiling a small module that transforms a function into WebAssembly.Function, but not sure we should go that way. This should be addressed, once the proposal lands to WebAssembly:

https://github.com/WebAssembly/js-types/blob/main/proposals/js-types/Overview.md#addition-of-webassemblyfunction

**Comment by LeXXik** [comment]
Actually the second crash might be related to instance reload, not sure. Didn't investigate deeply. I will create another issue.

**Comment by willeastcott** [comment]
Looked into this — I don't think there's an actual crash. The `TypeError` is caught internally by Emscripten and only surfaces because "pause on caught exceptions" is on; resuming continues normally.

`Ammo.addFunction` (Emscripten's helper for turning a JS function into a Wasm table pointer) uses a `try/catch` as feature-detection:

```js
try {
    table.set(slot, jsFunc);   // throws TypeError — a JS fn can't go in a funcref table
} catch (err) {
    if (!(err instanceof TypeError)) throw err;
    // fallback: WebAssembly.Function if available, otherwise compile a tiny
    // wrapper module (new WebAssembly.Module → Instance → exports.f)
    table.set(slot, wrappedWasmFn);
}
return slot;  // valid pointer either way
```

So the `TypeError` is expected and handled. Worth noting: the wrapper-module workaround you mentioned is exactly what the `catch` block already does — `WebAssembly.Function` is just an optimization used when present, not a requirement, so we aren't actually blocked on that proposal landing.

We also can't easily avoid the caught exception. The only alternative to `addFunction` is dropping `setInternalTickCallback` and checking collisions once per frame instead of after each internal substep — which is the missed-collision behavior #2074 fixed by introducing the tick callback in the first place.

Since the app runs correctly and the exception never escapes into engine code, I'd suggest closing this as expected Emscripten behavior — unless there's a repro where physics actually breaks rather than just the debugger pausing.

**Comment by LeXXik** [comment]
That is fine. @kungfooman did find an actual cause of the crash and fixed it. It was a while back, though, so can't recall what it was.

**Comment by willeastcott** [comment]
Ah interesting - good to know. 😄
